### PR TITLE
softmax for fp16 with fp32 accumulator

### PR DESCRIPTION
**Pull request by szha**
## Description ##
softmax for fp16 with fp32 accumulator

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] softmax for fp16 with fp32 accumulator

### CODE REVIEW

**Comment by szha** [comment]
@ptrendx @DickJC123 I'd appreciate your review on this. Thanks.

File: src/operator/nn/softmax-inl.h [hunk 1]
```diff
@@ -36,22 +36,22 @@ namespace op {
 namespace mxnet_op {
 
 struct softmax_fwd {
-  template<typename DType>
-  MSHADOW_XINLINE static DType Map(DType a, DType b) {
+  template<typename DType, typename AType>
+  MSHADOW_XINLINE static DType Map(DType a, AType b) {
     return DType(expf(a)/b);
```
Comment (ptrendx) [inline src/operator/nn/softmax-inl.h]: This still could overflow during cast to DType. I would say this function should return AType.

File: src/operator/nn/softmax-inl.h [hunk 2]
```diff
@@ -36,22 +36,22 @@ namespace op {
...
   }
 };
 
 
 struct log_softmax_fwd {
-  template<typename DType>
-  MSHADOW_XINLINE static DType Map(DType a, DType b) {
+  template<typename DType, typename AType>
+  MSHADOW_XINLINE static DType Map(DType a, AType b) {
     return DType(a - logf(b));
```
Comment (ptrendx) [inline src/operator/nn/softmax-inl.h]: Same here.

Comment (szha) [reply src/operator/nn/softmax-inl.h]: Thanks. Good point. The same should apply to the gradient kernels

**Comment by szha** [comment]
I don't think we can change the return type of log_softmax in minor release.

**Comment by szha** [comment]
I will add an optional dtype option for this.

**Comment by vandanavk** [comment]
@mxnet-label-bot add [Operator, pr-awaiting-review]

File: src/operator/mxnet_op.h [hunk 3]
```diff
@@ -249,6 +249,48 @@ inline int get_num_threads<cpu>(const int N) {
     LOG(FATAL) << "Unknown type enum " << type;            \
   }
 
+#define MXNET_REAL_ACC_TYPE_SWITCH(type, DType, AType, ...)\
+  switch (type) {                                          \
+  case mshadow::kFloat32:                                  \
+    {                                                      \
+      typedef float DType;                                 \
+      typedef float AType;                                 \
```
Comment (eric-haibin-lin) [inline src/operator/mxnet_op.h]: For fp32, should AType be double instead?

File: src/operator/nn/softmax-inl.h [hunk 4]
```diff
@@ -275,14 +292,70 @@ inline void SoftmaxGrad(Stream<gpu> *s, DType *out, DType *ograd,
...
 };
 
+inline bool SoftmaxOpType(const nnvm::NodeAttrs& attrs,
+                          std::vector<int>* in_attrs,
+                          std::vector<int>* out_attrs) {
+  CHECK_EQ(in_attrs->size(), 1);
+  CHECK_EQ(out_attrs->size(), 1);
+  const SoftmaxParam& param = nnvm::get<SoftmaxParam>(attrs.parsed);
+
+  int arg_dtype = param.dtype.has_value()?param.dtype.value():-1,
```
Comment (larroy) [inline src/operator/nn/softmax-inl.h]: can we add spaces around ternary for readability?

File: src/operator/nn/softmax-inl.h [hunk 5]
```diff
@@ -275,14 +292,70 @@ inline void SoftmaxGrad(Stream<gpu> *s, DType *out, DType *ograd,
...
+      in_dtype = (*in_attrs)[0],
```
Comment (larroy) [inline src/operator/nn/softmax-inl.h]: I would assign on separate expressions.

**Comment by szha** [comment]
@larroy thanks. Accuracy would be tied to a model and a dataset and is thus harder to reason. The more immediate effect is it's harder to overflow when accumulating e^x.

**Comment by szha** [comment]
I intend to change the PR so that GradUseInOut only happens when dtype is specified and different from input type.

File: src/operator/nn/softmax-inl.h [hunk 6]
```diff
@@ -275,11 +293,108 @@ inline void SoftmaxGrad(Stream<gpu> *s, DType *out, DType *ograd,
...
+  CHECK_EQ(out_attrs->size(), 1);
+  const SoftmaxParam& param = nnvm::get<SoftmaxParam>(attrs.parsed);
+
+  int arg_dtype = param.dtype.has_value() ? param.dtype.value() : -1;
+  int in_dtype = (*in_attrs)[0];
+  int out_dtype = (*out_attrs)[0];
+
+  if (out_dtype != -1 && in_dtype != -1) {
+    TYPE_ASSIGN_CHECK(*out_attrs, 0, arg_dtype);
+    TYPE_ASSIGN_CHECK(*in_attrs, 0, out_dtype);
```
Comment (eric-haibin-lin) [inline src/operator/nn/softmax-inl.h]: why force type assign here?

Comment (eric-haibin-lin) [reply src/operator/nn/softmax-inl.h]:
```
if (arg_dtype != -1) {
    TYPE_ASSIGN_CHECK(*out_attrs, 0, arg_dtype);
    // it's ok if type assign returns false when arg_dtype is specified.
    type_assign(in_attrs, out_attrs[0]);
    return true;
} else {
    type_assign(in_attrs, out_attrs[0]);
    return type_assign(out_attrs, in_attrs[0]);
}
```

Comment (szha) [reply src/operator/nn/softmax-inl.h]: 为啥后一个condition里不是两个check

File: src/operator/nn/softmax.cc [hunk 7]
```diff
@@ -102,15 +102,36 @@ Example::
...
+.set_attr<nnvm::FInplaceOption>("FInplaceOption",
+  [](const NodeAttrs& attrs){
+    return std::vector<std::pair<int, int> >{{0, 0}};
+  })
+.add_argument("data", "NDArray-or-Symbol", "The input array.")
 .add_arguments(SoftmaxParam::__FIELDS__());
 
-MXNET_OPERATOR_REGISTER_BINARY(_backward_softmax)
+NNVM_REGISTER_OP(_backward_softmax)
+.set_num_inputs(3)
```
Comment (eric-haibin-lin) [inline src/operator/nn/softmax.cc]: num inputs may be 2 or 3 depending on params.dtype. Reference impl: https://github.com/apache/incubator-mxnet/blob/master/src/operator/tensor/elemwise_sum.cc

File: src/operator/nn/softmax.cc [hunk 8]
```diff
@@ -102,15 +102,36 @@ Example::
...
+.set_num_outputs(1)
+.set_attr<nnvm::FListInputNames>("FListInputNames",
+  [](const NodeAttrs& attrs) {
+    return std::vector<std::string>{"ograd", "data", "output"};
+  })
+.set_attr<nnvm::FInferShape>("FInferShape", SoftmaxGradOpShape)
+.set_attr<nnvm::FInferType>("FInferType", SoftmaxGradOpType)
+.set_attr<nnvm::FInplaceOption>("FInplaceOption", SoftmaxGradOpInplaceOption)
+.add_argument("ograd", "NDArray-or-Symbol", "gradient of output")
```
Comment (eric-haibin-lin) [inline src/operator/nn/softmax.cc]: Add an argument with `NDArray-or_Symbol[]` should be sufficient. Reference impl: https://github.com/apache/incubator-mxnet/blob/master/src/operator/tensor/elemwise_sum.cc

**Comment by eric-haibin-lin** [comment]
LGTM. @ptrendx any other concerns?

**Pull request merged**